Move LiskHQ/lisk-build into LiskHQ/lisk and introduce basic Makefile - Closes #2400

fchavant opened the pull request:
### What was the problem?
The build script was separated from the code altough it's tightly coupled with its dependencies.

### How did I fix it?
Moved `build.sh` from LiskHQ/lisk-build into the `build` subdirectory.

### How to test it?
Follow the instructions in `build/README.md`; it boils down to using Ubuntu 16.04, installing the dependencies and running `make LISK_NET=<test|main>`.

### Review checklist

* The PR resolves #2400 
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

francoilol committed:
Move build from LiskHQ/lisk-build; add Makefile

Nazgolze committed:
Add python and wget to reqs in README.md

francoilol committed:
Read .build file

nazarhussain commented on build/README.md:
````diff
@@ -0,0 +1,19 @@
+### How-To
+
+```
+make LISK_NET=main
````
We use variable `LISK_NETWORK` all around the app, please use same here as well. Also use full network names.

fchavant replied:
`LISK_NETWORK` would contain `mainnet` while `LISK_NET` is set to `main`; this is because `build.sh` requires `main` or `test` to be passed as parameter.

nazarhussain commented on build/README.md:
````diff
@@ -0,0 +1,19 @@
+### How-To
+
+```
+make LISK_NET=main
````
I think better shell syntax is to put ENV variables before start of command. 

```
LISK_NETWORK=mainnet  make
```

fchavant replied:
That is how you pass a variable to make.

nazarhussain commented on build/build.sh:
```diff
@@ -0,0 +1,192 @@
+
+################################################################################
+
+# Requirements for Ubuntu 14.04.5 LTS base install:
```
In readme its specified `16.04 LTS`

nazarhussain commented on build/build.sh:
```diff
@@ -0,0 +1,192 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+######################################################################
```
We use different license header in Lisk Core, please copy same from any other file.

nazarhussain commented on build/build.sh:
```diff
@@ -0,0 +1,192 @@
+echo "Downloading jq..."
+echo "--------------------------------------------------------------------------"
+[[ -f "$JQ_FILE" ]] || wget -nv "$JQ_BIN_URL" --output-document="$JQ_FILE"
+echo "$JQ_SHA256SUM  $JQ_FILE" |sha256sum -c
```
`jq` is a dependency of the build process. Can we not just use the available jq binary and don't download new copy.

nazarhussain commented on build/config.sh:
```diff
@@ -0,0 +1,74 @@
+LISK_SCRIPTS_FILE="$LISK_SCRIPTS_DIR.tar.gz"
+LISK_SCRIPTS_URL="https://github.com/LiskHQ/lisk-scripts/archive/v$LISK_SCRIPTS_VERSION.tar.gz"
+
+NODE_VERSION="8.12.0"
```
Is it possible to extract this version from `package.json` key `engines.node`?

fchavant replied:
I was thinking about taking the version in `.nvmrc` actually: https://github.com/LiskHQ/lisk-build/issues/351

nazarhussain commented on build/build.sh:
```diff
@@ -0,0 +1,192 @@
+	"./bin/$JQ_FILE" '.fileLogLevel="info"' etc/snapshot.json |sponge etc/snapshot.json
+	"./bin/$JQ_FILE" '.db.database="lisk_snapshot"' etc/snapshot.json |sponge etc/snapshot.json
+	"./bin/$JQ_FILE" '.peers.list=[]' etc/snapshot.json |sponge etc/snapshot.json
+	"./bin/$JQ_FILE" '.loading.loadPerIteration=101' etc/snapshot.json |sponge etc/snapshot.json
```
I don't think hardcode the configuration logic in build process is a good idea. We should create a seperate json file for with these values and then copy/merge it here.

nazarhussain commented on build/build.sh:
```diff
@@ -0,0 +1,192 @@
+	echo "--------------------------------------------------------------------------"
+	echo "export LISK_NETWORK=${LISK_NETWORK}net" >>env.sh
+
+	echo "Create default custom config.json..."
```
> Create default custom config.json

default or custom?

fchavant replied:
The default version (empty: `{}`) of the custom configuration file.

nazarhussain commented on build/build.sh:
```diff
@@ -0,0 +1,192 @@
+	npm install --production --global "pm2@$PM2_VERSION"
+	npm install --production --global "lisk-commander@$LISK_COMMANDER_VERSION"
+
+	date "+%H:%M:%S %d/%m/%Y" >build
```
The file you referred in `app.js` is `.build`, I think `.` is missing here.

nazarhussain commented on build/Makefile:
```diff
@@ -0,0 +1,22 @@
+src/lisk-$(lisk_version).tgz: ../REVISION
+	cd ..; tar czf build/src/lisk-$(lisk_version).tgz \
+	    --exclude-vcs --exclude=build --exclude-ignore=build/build.ignore \
+	    --transform='s#^./#package/#' .
```
It seems you completely removed the behavior of `npm pack` with the file `build.ignore`. I feel it's not good practice. You should use `.npmignore` file as mostly developers will not aware of `build.ignore` file. For developer point of view they will always modify/look `.npmignore` to see which files will be shipped in release.

fchavant replied:
While we do not use the same tool (`tar` instead of `npm pack`) or ignore list the behaviour is the same: we create the same file in the same format . It is true that we ignore a few more files though.
The reason we do not use the existing `.npmignore` is because its format is slightly different (and not documented?) -- do we need it at all if we never call `npm pack`?

nazarhussain replied:
@fchavant That's not true exactly, Just check this out https://docs.npmjs.com/misc/developers

> .npmignore files follow the same pattern rules as .gitignore files:
> 
> Blank lines or lines starting with # are ignored.
> Standard glob patterns work.
> You can end patterns with a forward slash / to specify a directory.
> You can negate a pattern by starting it with an exclamation point !.
> .
> .
> Additionally, everything in node_modules is ignored, except for bundled dependencies. npm automatically handles this for you, so don’t bother adding node_modules to .npmignore.
> 

So `.npmignore` is very well documented and actually follow the exact rules of `.gitingore`. 

Now just consider a scenario that you ask any Core developer (any NodeJS developer):

- Which files will be shipped in production? 
- A developer created a file, but he don't want it to be shipped in production. 

Everyone will go and look for `.npmignore` file. So removing a file which have so much significance in technology domain is not right approach. 

The problem during the build is that you don't want to use `npm pack`, so you come up with a workaround for it, but by not removing `.npmignore` file.

fchavant replied:
So what is your suggestion?

nazarhussain replied:
Two possibilities. 

1. Use `npm pack` to we can use its power and also benefit if it introduce new feature in future. 
2 Use custom bash script to strip down the files using `.npmignore` and then pass it to tar.

shuse2 replied:
Now it's often mainly use 
https://docs.npmjs.com/files/package.json#files

nazarhussain commented on build/build.ignore:
```diff
@@ -0,0 +1,8 @@
+Gruntfile.js
+Jenkinsfile*
+tasks
+test
```
This file actually not behave same like `.npmignore`. Please see the list there. But better not to introduce new file and use only `.npmignore`.

francoilol committed:
Remove outdated comment from build.sh

francoilol committed:
Update license header

francoilol committed:
Correct typo

Nazgolze committed:
Use lisk-commander 2.0.0

francoilol committed:
Package PostgreSQL 10.6

francoilol committed:
Use LISK_NETWORK instead of LISK_NET

Nazgolze reviewed (dismissed):
lgtm

francoilol committed:
Use Node.JS 8.13.0

Nazgolze reviewed (changes requested):
I missed this before - there's no logs directory in the resulting tarball. After manually adding it in though, the build works

francoilol committed:
Create logs directory during build

francoilol committed:
Create archive in build.sh using downloaded npm

To use 'npm pack' we now:
 - download Node.JS first,
 - unpack and use it to call 'npm pack' and
 - continue with the build as before

MaciejBaj commented:
We decided that:
- `.npmignore` is used only for build
- we should use .npmignore to create builds instead of .buildignore
- .buildignore can be removed as it's not used

MaciejBaj commented on build/config.sh:
```diff
@@ -0,0 +1,69 @@
+LISK_SCRIPTS_FILE="$LISK_SCRIPTS_DIR.tar.gz"
+LISK_SCRIPTS_URL="https://github.com/LiskHQ/lisk-scripts/archive/v$LISK_SCRIPTS_VERSION.tar.gz"
+
+NODE_VERSION="8.13.0"
```
I will create a PR to update the node version to 8.13.0 for the sources as well.
